feat(dpf): support per-deployment mandatory services override (#3203)
NVIDIA/bare-metal-manager-core: docs/manuals/dpf.md

docker_image_pull_secret = "dpf-pull-secret"
```

> #### Per-deployment configuration (`[dpf.deployments.*]`)

Reviewer: H4 and lower don't stand out a lot, visually, on the page, and don't show up in the right nav. I think that's probably ok here, but just wanted you to be aware of it.

Each DPU generation is provisioned by its own `DPUDeployment`, configured under
`[dpf.deployments.<name>]`. **BF3** is always present with built-in defaults;
**BF4 (generic)** is opt-in and is activated only when a
`[dpf.deployments.bf4_generic]` table is present. Both deployments run
side-by-side, each with its own BFB, `DPUFlavor`, and `DPUDeployment`.

Every active deployment must have a **unique** `deployment_name`, `flavor_name`,
and `node_label_key`; carbide-api validates this at startup and refuses to start

> if any two deployments collide.

Reviewer: I know that _any_ two deployments colliding also takes care of the case where _more than_ two collide, but this makes it simpler and clearer, avoiding questions like "but what if I have five?"

```suggestion
if any deployments collide.
```

Author: fixed.

```toml

> # BF3 — present by default. Override only if any change is needed.

Reviewer: ```suggestion
# BF3 is present by default. Override only if any change is needed.
```

Author: fixed.

[dpf.deployments.bf3]
bfb_url         = "https://content.mellanox.com/BlueField/BFBs/Ubuntu24.04/bf-bundle-3.2.2-125_26.02_ubuntu-24.04_64k_prod.bfb"
flavor_name     = "carbide-dpu-flavor"
deployment_name = "nico-deployment-v2"
node_label_key  = "carbide.nvidia.com/controlled.node.v2"

# BF4 generic — opt-in. Adding this table provisions BF4 DPUs via a second

> # DPUDeployment alongside BF3. All identifiers must differ from bf3's.

Reviewer: ```suggestion
# BF4 generic is opt-in. Add this table to provision BF4 DPUs via a second
# DPUDeployment alongside BF3. All identifiers must differ from BF3's.
```

Author: fixed.

deployment_name = "nico-deployment-bf4"
node_label_key  = "carbide.nvidia.com/controlled.node.bf4"
```

Per-deployment field reference:

| TOML key | Required | Default (bf3) | Meaning |
| --- | :---: | --- | --- |
| `bfb_url` | no | BF3 bf-bundle URL | BlueField firmware bundle (BFB) used to provision the DPU. |
| `flavor_name` | yes | `carbide-dpu-flavor` | `DPUFlavor` CR name for this deployment. |
| `deployment_name` | yes | `nico-deployment-v2` | `DPUDeployment` CR name. |
| `node_label_key` | yes | `carbide.nvidia.com/controlled.node.v2` | Node-selector label key applied to this deployment's DPUNodes. |
| `services` | no | inherit `[dpf.services]` | Optional per-deployment mandatory-services override (see below). |

**Per-deployment services override.** By default every deployment inherits the
top-level `[dpf.services]` mandatory services. A deployment can pin its own
versions by adding a `[dpf.deployments.<name>.services]` block with the same six
sub-tables as `[dpf.services]` (`dts`, `doca_hbn`, `dpu_agent`, `dhcp_server`,
`fmds`, `otel`). This override **replaces** the inherited set for that
deployment; any service sub-table you omit falls back to its **built-in

> default**, *not* to the top-level `[dpf.services]` value — so specify all six

Reviewer: ```suggestion
deployment; any service sub-table you omit falls back to its **built-in
default**, *not* to the top-level `[dpf.services]` value, so specify all six
```

Author: fixed.

Field reference (all under `[dpf]`):

| TOML key | Type | Default | Meaning |
| --- | --- | --- | --- |
| `enabled` | bool | `false` | Master switch. Must be `true` to use DPF-based provisioning. |

> | `docker_image_pull_secret` | string | `dpf-pull-secret` | Pull Secret applied to every mandatory service except `dts`/`doca_hbn`. |

Reviewer: ```suggestion
| `docker_image_pull_secret` | string | `dpf-pull-secret` | Pull Secret applied to every mandatory service except `dts` and `doca_hbn`. |
```

Author: fixed.

DPF initialization objects are created in the host cluster.

On startup with `[dpf].enabled = true`, carbide-api creates the following
objects in the `dpf-operator-system` namespace. It does this **once per active

> deployment** under `[dpf.deployments.*]` (BF3 always, plus `bf4_generic` when

Reviewer: ```suggestion
objects in the `dpf-operator-system` namespace. It does this **once for each active
deployment** in `[dpf.deployments.*]` (BF3 always, plus `bf4_generic` when
```

Author: fixed.

that table is present), using that deployment's own `bfb_url`, `flavor_name`,
and `deployment_name`:

- a `Secret` (`bmc-shared-password`) holding the shared BMC password (shared
  across deployments),
- a `BFB` CR named `bf-bundle-<sha256(bfb_url)>`, from the deployment's `bfb_url`,
- a `DPUFlavor` CR named `<flavor_name>-<spec-hash>` (the 16-character hex suffix is a SHA-256 digest of the spec, so any change to the flavor — including adding or changing `[dpf.proxy]` — produces a new name and triggers reprovisioning of that deployment's DPUs),
- a set of `DPUServiceInterface`, `DPUServiceTemplate`,
  `DPUServiceConfiguration`, and `DPUServiceNAD` CRs — one per mandatory

>   DPUService (`dts`, `doca-hbn`, `carbide-dpu-agent`, `carbide-dhcp-server`,

Reviewer: (Continues in the next comment)

```suggestion
- A `Secret` (`bmc-shared-password`) holding the shared BMC password (shared
  across deployments)
- A `BFB` CR named `bf-bundle-<sha256(bfb_url)>`, from the deployment's `bfb_url`
- A `DPUFlavor` CR named `<flavor_name>-<spec-hash>`. (The 16-character hex suffix is a SHA-256 digest of the spec. Any change to the flavor, including adding or changing `[dpf.proxy]`, produces a new name and triggers reprovisioning of that deployment's DPUs).)
- A set of `DPUServiceInterface`, `DPUServiceTemplate`,
`DPUServiceConfiguration`, and `DPUServiceNAD` CRs, one per mandatory
DPUService (`dts`, `doca-hbn`, `carbide-dpu-agent`, `carbide-dhcp-server`, `carbide-fmds`, and `carbide-otelcol`). The set is built from the deployment's resolved services -- either its `[dpf.deployments.<name>.services]` override if set, otherwise the top-level `[dpf.services]`.
```

Author: fixed.

  `carbide-fmds`, `carbide-otelcol`), built from the deployment's resolved
  services (its `[dpf.deployments.<name>.services]` override if set, otherwise
  the top-level `[dpf.services]`),
- a `DPUDeployment` CR named after the deployment's `deployment_name`, which
  references the BFB, the DPUFlavor, and the service templates above, and which
  the DPF operator then reconciles into actual `DPUService` and per-DPU

>   resources.

Reviewer: (Continued from previous)

```suggestion
- A `DPUDeployment` CR named after the deployment's `deployment_name`, which
  references the BFB, the DPUFlavor, and the service templates above, and which
  the DPF operator then reconciles into actual `DPUService` and per-DPU
  resources.
```

Author: fixed.
